sassoftware/vscode-sas-extension
# feat: username and password support for SSH connection type

**Summary**
- Adds support for username and password to the 9.4 Remote (SSH) connection type.
- Refactor SSH auth process to use a handler that can support use of agent, manually specifying a private keyfile (with or without a passphrase) in the connection profile,  using a traditional username password, or using keyboard-interactive mode, depending on how the ssh server is configured.
- Run initialization code to set working dir to the WORK lib path to autocleanup ODS results on session close
- Add keepalive configuration to ssh client to prevent idle disconnects for long-running programs



Todos:
- [x] Use authHandler to handle authentication requests. See [ssh2](https://github.com/mscdex/ssh2) docs for authHandler support. See prototype by @chunky [here](https://github.com/sassoftware/vscode-sas-extension/commit/d276e0c27026fa18afb9ebbb1fb61824016af10d). We need to give some thought on how we define what supported auth methods should be attempted. I think it would be great to let the server tell us what the supported methods are, and then react to them. Kerberos methods are going to be a problem because the ssh2 library cant support those auth methods currently. For each supported method, the auth should be attempted, moving onto the next one during a failure to avoid issues like [this](https://github.com/sassoftware/vscode-sas-extension/issues/1091#issuecomment-2225596787). 
- [x] Username/Password Support. Prompt for password "Just in time" on session establishment. ~~Store password in Secrets API so that the user is not prompted for a password on each session establishment after initial profile creation.~~ Use masked input field like we do for IOM on the text input prompt.
- [x] Keyboard Interactive support. Similar to, but not the same as password. MFA scenarios will request this auth method. We need to detect the incoming requests for input and prompt the user for answers, before sending on the answers to the auth process.
- [x] Maintain support for agent based authentication for users that want passwordless or agent based setup. Users should still be able to specify an identify file for keys that have a passphrase.
- [x] The `setTimeout` logic will need to be adjusted so that we dont errantly timeout while a user is typing in a password or passphrase.
- [x] Add username to password prompt text 
- [x] Add vscode metadata for new privateKeyFilePath profile field
- [x] Add support for work directory detection for #1176 
- [x] Add keepalives and max session time config options for #1156
- [x] Doc updates

**Testing**

Connection Profile Prompts:
- Create new connection profile should offer private key path as an optional prompt and create the profile with the value that the user specifies

Auth Refactor:
Users can successfully authenticate to the server to run sas programs using the following auth methods:
- SSH Agent (See current doc)
- Private key specified  in privateKeyFilePath field in connection profile (no passphrase)
- Public key specified in privateKeyFilePath field in connection profile (with a passphrase)
- Using a traditional username/password
- Using Keyboard Interactive (RHEL disables this by default in modern versions, will need to enable it in sshd config on the server and set the AuthenticationMethods field such that this method is used instead of password)

Work Directory Detection
- ODS SAS result html files are no longer written out to the users home directory on the SSH Server
- Verify that the SAS Log Output contains an output entry for the working dir getting set to the WORK dir generated.
- User can change the work directory by passing -WORK [path] option as a sasOption on the connection profile options array

Keepalive Changes:
- Verify that long running program (using sleep function) does not result in the ssh server disconnecting the client
